File: packages/provide-uterm-go/vt/stream.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package vt

import "strings"

// Control characters recognized by the stream (pyte.control).
const (
	ctrlNUL   = 0x00
	ctrlBEL   = 0x07
	ctrlBS    = 0x08
	ctrlHT    = 0x09
	ctrlLF    = 0x0a
	ctrlVT    = 0x0b
	ctrlFF    = 0x0c
	ctrlCR    = 0x0d
	ctrlSO    = 0x0e
	ctrlSI    = 0x0f
	ctrlCAN   = 0x18
	ctrlSUB   = 0x1a
	ctrlESC   = 0x1b
	ctrlDEL   = 0x7f
	ctrlCSIC1 = 0x9b
	ctrlSTC1  = 0x9c
	ctrlOSCC1 = 0x9d
)

// Parser states.
const (
	stGround = iota
	stEscape
	stSharp
	stPercent
	stCharset
	stCSI
	stCSIDollar
	stOSCCode
	stOSCParam
)

// Stream is a state machine that parses a stream of terminal input and
// dispatches events to a Screen, mirroring pyte.streams.Stream (the
// text-based stream).
type Stream struct {
	// UseUTF8 mirrors pyte.Stream.use_utf8 (default true): while set,
	// SO/SI and ESC ( / ESC ) charset selection are ignored.
	UseUTF8 bool

	screen          *Screen
	takingPlainText bool
	state           int

	// CSI collection state.
	csiParams  []int
	csiCurrent int
	csiPrivate bool

	// OSC collection state.
	oscCode  rune
	oscParam strings.Builder
	oscESC   bool

	// Charset designation state (ESC ( or ESC )).
	charsetMode rune
}

// NewStream creates a stream dispatching to the given screen.
func NewStream(screen *Screen) *Stream {
	return &Stream{
		UseUTF8:         true,
		screen:          screen,
		takingPlainText: true,
	}
}

// isSpecial reports whether r must go through the state machine rather
// than being drawn as part of a plain-text run.
func isSpecial(r rune) bool {
	return r == ctrlNUL || (r >= ctrlBEL && r <= ctrlSI) ||
		r == ctrlESC || r == ctrlDEL || r == ctrlCSIC1 || r == ctrlOSCC1
}

// Feed consumes a chunk of input, advancing parser state as necessary.
// Maximal runs of plain text are dispatched to Screen.Draw as single
// batches, exactly like pyte's regex-based fast path — batch boundaries
// are observable because Draw stops at unprintable characters.
func (st *Stream) Feed(data string) {
	runes := []rune(data)
	length := len(runes)
	offset := 0
	for offset < length {
		if st.takingPlainText {
			end := offset
			for end < length && !isSpecial(runes[end]) {
				end++
			}
			if end > offset {
				st.screen.Draw(string(runes[offset:end]))
				offset = end
			} else {
				st.takingPlainText = false
			}
		} else {
			st.takingPlainText = st.send(runes[offset])
			offset++
		}
	}
}

// send advances the state machine by one rune, reporting whether the
// parser is back at ground (allowing the plain-text fast path).
func (st *Stream) send(r rune) bool {
	switch st.state {
	case stGround:
		st.ground(r)
	case stEscape:
		st.escape(r)
	case stSharp:
		if r == '8' {
			st.screen.AlignmentDisplay()
		}
		st.state = stGround
	case stPercent:
		// ESC % — select-other-charset is a noop for text streams.
		st.state = stGround
	case stCharset:
		if !st.UseUTF8 {
			st.screen.DefineCharset(string(r), string(st.charsetMode))
		}
		st.state = stGround
	case stCSI:
		st.csi(r)
	case stCSIDollar:
		// XTerm ESC ] ... $ [a-z] sequences: skip one char and bail.
		st.state = stGround
	case stOSCCode:
		st.oscBegin(r)
	case stOSCParam:
		st.osc(r)
	}
	return st.state == stGround
}

// ground handles a special character in the ground state.
func (st *Stream) ground(r rune) {
	switch {
	case r == ctrlESC:
		st.state = stEscape
	case isBasic(r):
		// Shifts are ignored in UTF-8 mode.
		if (r == ctrlSI || r == ctrlSO) && st.UseUTF8 {
			return
		}
		st.dispatchBasic(r)
	case r == ctrlCSIC1:
		st.enterCSI()
	case r == ctrlOSCC1:
		st.state = stOSCCode
	case r == ctrlNUL || r == ctrlDEL:
		// Ignored.
	default:
		st.screen.Draw(string(r))
	}
}

// escape handles the character following ESC.
func (st *Stream) escape(r rune) {
	switch r {
	case '[':
		st.enterCSI()
	case ']':
		st.state = stOSCCode
	case '#':
		st.state = stSharp
	case '%':
		st.state = stPercent
	case '(', ')':
		st.charsetMode = r
		st.state = stCharset
	default:
		st.dispatchEscape(r)
		st.state = stGround
	}
}

// enterCSI resets CSI collection state and enters the CSI state.
func (st *Stream) enterCSI() {
	st.csiParams = st.csiParams[:0]
	st.csiCurrent = 0
	st.csiPrivate = false
	st.state = stCSI
}

// csi handles one character of a control sequence.
func (st *Stream) csi(r rune) {
	switch {
	case r == '?':
		st.csiPrivate = true
	case r == ctrlBEL || r == ctrlBS || r == ctrlHT || r == ctrlLF ||
		r == ctrlVT || r == ctrlFF || r == ctrlCR:
		// Control characters allowed (and executed) inside CSI.
		st.dispatchBasic(r)
	case r == ' ' || r == '>':
		// Secondary DA is not supported.
	case r == ctrlCAN || r == ctrlSUB:
		// CAN/SUB abort the sequence; the substitute character is drawn.
		st.screen.Draw(string(r))
		st.state = stGround
	case r >= '0' && r <= '9':
		if st.csiCurrent < 100000 { // Parameters saturate at 9999 anyway.
			st.csiCurrent = st.csiCurrent*10 + int(r-'0')
		}
	case r == '$':
		// XTerm-specific sequences are not supported: skip one char.
		st.state = stCSIDollar
	default:
		st.csiParams = append(st.csiParams, min(st.csiCurrent, 9999))
		if r == ';' {
			st.csiCurrent = 0
		} else {
			st.dispatchCSI(r, st.csiParams, st.csiPrivate)
			st.state = stGround
		}
	}
}

// oscBegin handles the character after ESC ] — the OSC command code.
func (st *Stream) oscBegin(r rune) {
	if r == 'R' || r == 'P' {
		// Reset/set palette: not implemented.
		st.state = stGround
		return
	}
	st.oscCode = r
	st.oscParam.Reset()
	st.oscESC = false
	st.state = stOSCParam
}

// osc accumulates the OSC parameter until a terminator (BEL, ST).
func (st *Stream) osc(r rune) {
	if st.oscESC {
		st.oscESC = false
		if r == '\\' { // ESC \ is the C0 string terminator.
			st.oscDispatch()
			return
		}
		st.oscParam.WriteRune(ctrlESC)
		st.oscParam.WriteRune(r)
		return
	}
	switch r {
	case ctrlESC:
		st.oscESC = true
	case ctrlSTC1, ctrlBEL:
		st.oscDispatch()
	default:
		st.oscParam.WriteRune(r)
	}
}

// oscDispatch fires the collected OSC command and returns to ground.
func (st *Stream) oscDispatch() {
	param := st.oscParam.String()
	// Drop the leading ";" separator (pyte drops the first code point
	// unconditionally).
	if _, size := firstRune(param); size > 0 {
		param = param[size:]
	}
	if st.oscCode == '0' || st.oscCode == '1' {
		st.screen.SetIconName(param)
	}
	if st.oscCode == '0' || st.oscCode == '2' {
		st.screen.SetTitle(param)
	}
	st.state = stGround
}

// firstRune returns the first rune of s and its encoded size (0 if empty).
func firstRune(s string) (rune, int) {
	for _, r := range s {
		return r, len(string(r))
	}
	return 0, 0
}

// isBasic reports whether r is a basic control character with a direct
// screen handler.
func isBasic(r rune) bool {
	return r >= ctrlBEL && r <= ctrlSI
}

// dispatchBasic executes a basic control character.
func (st *Stream) dispatchBasic(r rune) {
	switch r {
	case ctrlBEL:
		st.screen.Bell()
	case ctrlBS:
		st.screen.Backspace()
	case ctrlHT:
		st.screen.Tab()
	case ctrlLF, ctrlVT, ctrlFF:
		st.screen.LineFeed()
	case ctrlCR:
		st.screen.CarriageReturn()
	case ctrlSO:
		st.screen.ShiftOut()
	case ctrlSI:
		st.screen.ShiftIn()
	}
}

// dispatchEscape executes a non-CSI escape sequence; unknown sequences
// are ignored (pyte routes them to a debug noop).
func (st *Stream) dispatchEscape(r rune) {
	switch r {
	case 'c': // RIS.
		st.screen.Reset()
	case 'D': // IND.
		st.screen.Index()
	case 'E': // NEL.
		st.screen.LineFeed()
	case 'H': // HTS.
		st.screen.SetTabStop()
	case 'M': // RI.
		st.screen.ReverseIndex()
	case '7': // DECSC.
		st.screen.SaveCursor()
	case '8': // DECRC.
		st.screen.RestoreCursor()
	}
}

// dispatchCSI executes a CSI sequence; unknown final characters are
// ignored (pyte routes them to a debug noop). Surplus parameters are
// discarded, missing ones default to 0.
func (st *Stream) dispatchCSI(final rune, params []int, private bool) {
	p := func(i int) int {
		if i < len(params) {
			return params[i]
		}
		return 0
	}
	scr := st.screen
	switch final {
	case '@': // ICH.
		scr.InsertCharacters(p(0))
	case 'A': // CUU.
		scr.CursorUp(p(0))
	case 'B', 'e': // CUD, VPR.
		scr.CursorDown(p(0))
	case 'C', 'a': // CUF, HPR.
		scr.CursorForward(p(0))
	case 'D': // CUB.
		scr.CursorBack(p(0))
	case 'E': // CNL.
		scr.CursorDown1(p(0))
	case 'F': // CPL.
		scr.CursorUp1(p(0))
	case 'G', '\'': // CHA, HPA.
		scr.CursorToColumn(p(0))
	case 'H', 'f': // CUP, HVP.
		scr.CursorPosition(p(0), p(1))
	case 'J': // ED.
		scr.EraseInDisplay(p(0))
	case 'K': // EL.
		scr.EraseInLine(p(0), private)
	case 'L': // IL.
		scr.InsertLines(p(0))
	case 'M': // DL.
		scr.DeleteLines(p(0))
	case 'P': // DCH.
		scr.DeleteCharacters(p(0))
	case 'X': // ECH.
		scr.EraseCharacters(p(0))
	case 'c': // DA.
		scr.ReportDeviceAttributes(p(0), private)
	case 'd': // VPA.
		scr.CursorToLine(p(0))
	case 'g': // TBC.
		scr.ClearTabStop(p(0))
	case 'h': // SM.
		scr.SetMode(private, params...)
	case 'l': // RM.
		scr.ResetMode(private, params...)
	case 'm': // SGR.
		scr.SelectGraphicRendition(params...)
	case 'n': // DSR.
		scr.ReportDeviceStatus(p(0))
	case 'r': // DECSTBM.
		scr.SetMargins(params...)
	}
}

File: packages/provide-uterm-csharp/src/Provide.Uterm/Vt/Stream.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

using System.Text;

namespace Provide.Uterm.Vt;

/// <summary>
/// State machine that parses terminal input and dispatches to a Screen (pyte Stream).
/// </summary>
public sealed class VtStream
{
    private const int CtrlNul = 0x00;
    private const int CtrlBel = 0x07;
    private const int CtrlBs = 0x08;
    private const int CtrlHt = 0x09;
    private const int CtrlLf = 0x0a;
    private const int CtrlVt = 0x0b;
    private const int CtrlFf = 0x0c;
    private const int CtrlCr = 0x0d;
    private const int CtrlSo = 0x0e;
    private const int CtrlSi = 0x0f;
    private const int CtrlCan = 0x18;
    private const int CtrlSub = 0x1a;
    private const int CtrlEsc = 0x1b;
    private const int CtrlDel = 0x7f;
    private const int CtrlCsiC1 = 0x9b;
    private const int CtrlStC1 = 0x9c;
    private const int CtrlOscC1 = 0x9d;

    private const int StGround = 0;
    private const int StEscape = 1;
    private const int StSharp = 2;
    private const int StPercent = 3;
    private const int StCharset = 4;
    private const int StCsi = 5;
    private const int StCsiDollar = 6;
    private const int StOscCode = 7;
    private const int StOscParam = 8;

    public bool UseUtf8 { get; set; } = true;

    private readonly Screen _screen;
    private bool _takingPlainText = true;
    private int _state;

    private readonly List<int> _csiParams = new();
    private int _csiCurrent;
    private bool _csiPrivate;

    private int _oscCode;
    private readonly StringBuilder _oscParam = new();
    private bool _oscEsc;
    private int _charsetMode;

    public VtStream(Screen screen) => _screen = screen;

    private static bool IsSpecial(int r) =>
        r == CtrlNul || (r >= CtrlBel && r <= CtrlSi) ||
        r == CtrlEsc || r == CtrlDel || r == CtrlCsiC1 || r == CtrlOscC1;

    public void Feed(string data)
    {
        var runes = data.EnumerateRunes().Select(r => r.Value).ToArray();
        var length = runes.Length;
        var offset = 0;
        while (offset < length)
        {
            if (_takingPlainText)
            {
                var end = offset;
                while (end < length && !IsSpecial(runes[end]))
                {
                    end++;
                }

                if (end > offset)
                {
                    _screen.Draw(string.Concat(runes[offset..end].Select(r => char.ConvertFromUtf32(r))));
                    offset = end;
                }
                else
                {
                    _takingPlainText = false;
                }
            }
            else
            {
                _takingPlainText = Send(runes[offset]);
                offset++;
            }
        }
    }

    private bool Send(int r)
    {
        switch (_state)
        {
            case StGround:
                Ground(r);
                break;
            case StEscape:
                Escape(r);
                break;
            case StSharp:
                if (r == '8')
                {
                    _screen.AlignmentDisplay();
                }

                _state = StGround;
                break;
            case StPercent:
                _state = StGround;
                break;
            case StCharset:
                if (!UseUtf8)
                {
                    _screen.DefineCharset(char.ConvertFromUtf32(r), char.ConvertFromUtf32(_charsetMode));
                }

                _state = StGround;
                break;
            case StCsi:
                Csi(r);
                break;
            case StCsiDollar:
                _state = StGround;
                break;
            case StOscCode:
                OscBegin(r);
                break;
            case StOscParam:
                Osc(r);
                break;
        }

        return _state == StGround;
    }

    private void Ground(int r)
    {
        if (r == CtrlEsc)
        {
            _state = StEscape;
        }
        else if (IsBasic(r))
        {
            if ((r == CtrlSi || r == CtrlSo) && UseUtf8)
            {
                return;
            }

            DispatchBasic(r);
        }
        else if (r == CtrlCsiC1)
        {
            EnterCsi();
        }
        else if (r == CtrlOscC1)
        {
            _state = StOscCode;
        }
        else if (r is CtrlNul or CtrlDel)
        {
            // ignored
        }
        else
        {
            _screen.Draw(char.ConvertFromUtf32(r));
        }
    }

    private void Escape(int r)
    {
        switch (r)
        {
            case '[':
                EnterCsi();
                break;
            case ']':
                _state = StOscCode;
                break;
            case '#':
                _state = StSharp;
                break;
            case '%':
                _state = StPercent;
                break;
            case '(':
            case ')':
                _charsetMode = r;
                _state = StCharset;
                break;
            default:
                DispatchEscape(r);
                _state = StGround;
                break;
        }
    }

    private void EnterCsi()
    {
        _csiParams.Clear();
        _csiCurrent = 0;
        _csiPrivate = false;
        _state = StCsi;
    }

    private void Csi(int r)
    {
        if (r == '?')
        {
            _csiPrivate = true;
        }
        else if (r is CtrlBel or CtrlBs or CtrlHt or CtrlLf or CtrlVt or CtrlFf or CtrlCr)
        {
            DispatchBasic(r);
        }
        else if (r is ' ' or '>')
        {
            // Secondary DA not supported
        }
        else if (r is CtrlCan or CtrlSub)
        {
            _screen.Draw(char.ConvertFromUtf32(r));
            _state = StGround;
        }
        else if (r >= '0' && r <= '9')
        {
            if (_csiCurrent < 100000)
            {
                _csiCurrent = _csiCurrent * 10 + (r - '0');
            }
        }
        else if (r == '$')
        {
            _state = StCsiDollar;
        }
        else
        {
            _csiParams.Add(Math.Min(_csiCurrent, 9999));
            if (r == ';')
            {
                _csiCurrent = 0;
            }
            else
            {
                DispatchCsi(r, _csiParams.ToArray(), _csiPrivate);
                _state = StGround;
            }
        }
    }

    private void OscBegin(int r)
    {
        if (r is 'R' or 'P')
        {
            _state = StGround;
            return;
        }

        _oscCode = r;
        _oscParam.Clear();
        _oscEsc = false;
        _state = StOscParam;
    }

    private void Osc(int r)
    {
        if (_oscEsc)
        {
            _oscEsc = false;
            if (r == '\\')
            {
                OscDispatch();
                return;
            }

            _oscParam.Append((char)CtrlEsc);
            _oscParam.Append(char.ConvertFromUtf32(r));
            return;
        }

        switch (r)
        {
            case CtrlEsc:
                _oscEsc = true;
                break;
            case CtrlStC1:
            case CtrlBel:
                OscDispatch();
                break;
            default:
                _oscParam.Append(char.ConvertFromUtf32(r));
                break;
        }
    }

    private void OscDispatch()
    {
        var param = _oscParam.ToString();
        if (param.Length > 0)
        {
            var first = param.EnumerateRunes().First();
            param = param[first.Utf16SequenceLength..];
        }

        if (_oscCode is '0' or '1')
        {
            _screen.SetIconName(param);
        }

        if (_oscCode is '0' or '2')
        {
            _screen.SetTitle(param);
        }

        _state = StGround;
    }

    private static bool IsBasic(int r) => r >= CtrlBel && r <= CtrlSi;

    private void DispatchBasic(int r)
    {
        switch (r)
        {
            case CtrlBel:
                _screen.Bell();
                break;
            case CtrlBs:
                _screen.Backspace();
                break;
            case CtrlHt:
                _screen.Tab();
                break;
            case CtrlLf:
            case CtrlVt:
            case CtrlFf:
                _screen.LineFeed();
                break;
            case CtrlCr:
                _screen.CarriageReturn();
                break;
            case CtrlSo:
                _screen.ShiftOut();
                break;
            case CtrlSi:
                _screen.ShiftIn();
                break;
        }
    }

    private void DispatchEscape(int r)
    {
        switch (r)
        {
            case 'c':
                _screen.Reset();
                break;
            case 'D':
                _screen.Index();
                break;
            case 'E':
                _screen.LineFeed();
                break;
            case 'H':
                _screen.SetTabStop();
                break;
            case 'M':
                _screen.ReverseIndex();
                break;
            case '7':
                _screen.SaveCursor();
                break;
            case '8':
                _screen.RestoreCursor();
                break;
        }
    }

    private void DispatchCsi(int final, int[] parameters, bool privateMode)
    {
        int P(int i) => i < parameters.Length ? parameters[i] : 0;
        var scr = _screen;
        switch (final)
        {
            case '@':
                scr.InsertCharacters(P(0));
                break;
            case 'A':
                scr.CursorUp(P(0));
                break;
            case 'B':
            case 'e':
                scr.CursorDown(P(0));
                break;
            case 'C':
            case 'a':
                scr.CursorForward(P(0));
                break;
            case 'D':
                scr.CursorBack(P(0));
                break;
            case 'E':
                scr.CursorDown1(P(0));
                break;
            case 'F':
                scr.CursorUp1(P(0));
                break;
            case 'G':
            case '`':
                scr.CursorToColumn(P(0));
                break;
            case 'H':
            case 'f':
                scr.CursorPosition(P(0), P(1));
                break;
            case 'J':
                scr.EraseInDisplay(P(0));
                break;
            case 'K':
                scr.EraseInLine(P(0), privateMode);
                break;
            case 'L':
                scr.InsertLines(P(0));
                break;
            case 'M':
                scr.DeleteLines(P(0));
                break;
            case 'P':
                scr.DeleteCharacters(P(0));
                break;
            case 'X':
                scr.EraseCharacters(P(0));
                break;
            case 'c':
                scr.ReportDeviceAttributes(P(0), privateMode);
                break;
            case 'd':
                scr.CursorToLine(P(0));
                break;
            case 'g':
                scr.ClearTabStop(P(0));
                break;
            case 'h':
                scr.SetMode(privateMode, parameters);
                break;
            case 'l':
                scr.ResetMode(privateMode, parameters);
                break;
            case 'm':
                scr.SelectGraphicRendition(parameters);
                break;
            case 'n':
                scr.ReportDeviceStatus(P(0));
                break;
            case 'r':
                scr.SetMargins(parameters);
                break;
        }
    }
}
